data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java

```diff
@@ -0,0 +1,147 @@
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.common.document.DocumentField;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Stream;
+
+@DataPrepperPlugin(name = "otel_trace_group_prepper", type = PluginType.PREPPER)
+public class OTelTraceGroupPrepper extends AbstractPrepper<Record<String>, Record<String>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(OTelTraceGroupPrepper.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final TypeReference<Map<String, Object>> MAP_TYPE_REFERENCE = new TypeReference<Map<String, Object>>() {};
+
+    private final OTelTraceGroupPrepperConfig otelTraceGroupPrepperConfig;
+    private final RestHighLevelClient restHighLevelClient;
+    // TODO: add metrics
+
+    public OTelTraceGroupPrepper(final PluginSetting pluginSetting) {
+        super(pluginSetting);
+        otelTraceGroupPrepperConfig = OTelTraceGroupPrepperConfig.buildConfig(pluginSetting);
+        restHighLevelClient = otelTraceGroupPrepperConfig.getEsConnectionConfig().createClient();
+    }
+
+    @Override
+    public Collection<Record<String>> doExecute(final Collection<Record<String>> rawSpanStringRecords) {
+        final List<Record<String>> recordsOut = new LinkedList<>();
+        final Map<Record<String>, Map<String, Object>> recordMissingTraceGroupToRawSpanMap = new HashMap<>();
+        final Set<String> traceIdsToLookUp = new HashSet<>();
+        for (Record<String> record: rawSpanStringRecords) {
+            try {
+                final Map<String, Object> rawSpanMap = OBJECT_MAPPER.readValue(record.getData(), MAP_TYPE_REFERENCE);
+                final String traceGroup = (String) rawSpanMap.get(OTelTraceGroupPrepperConfig.TRACE_GROUP_FIELD);
+                final String traceId = (String) rawSpanMap.get(OTelTraceGroupPrepperConfig.TRACE_ID_FIELD);
+                if (traceGroup == null || traceGroup.equals("")) {
+                    traceIdsToLookUp.add(traceId);
+                    recordMissingTraceGroupToRawSpanMap.put(record, rawSpanMap);
+                } else {
+                    recordsOut.add(record);
+                }
+            } catch (JsonProcessingException e) {
+                recordsOut.add(record);
+                LOG.error("Failed to parse the record: [{}]", record.getData());
```

Comment:
nit - Should we even add the record to the output if it can't be parsed?

Reply:
Good point. That will lead to propagate the error to sink. I will remove L69.